realm/realm-library/src/main/java/io/realm/RealmCache.java

```diff
@@ -0,0 +1,271 @@
+     * Realm.
+     *
+     * @throws IllegalArgumentException If the new configuration isn't valid.
+     */
+    private static void validateAgainstExistingConfiguration(RealmConfiguration newConfiguration,
+                                                             RealmConfiguration cachedConfiguration) {
+        // Check that encryption keys aren't different
+        if (!Arrays.equals(cachedConfiguration.getEncryptionKey(), newConfiguration.getEncryptionKey())) {
+            throw new IllegalArgumentException(DIFFERENT_KEY_MESSAGE);
+        }
+
+        // Check schema versions are the same
+        if (cachedConfiguration.getSchemaVersion() != newConfiguration.getSchemaVersion()) {
+            throw new IllegalArgumentException(String.format("Configurations cannot have different schema versions " +
+                            "if used to open the same file. %d vs. %d", cachedConfiguration.getSchemaVersion(),
+                    newConfiguration.getSchemaVersion()));
+        }
+
+        // Check that schema is the same
+        RealmProxyMediator cachedSchema = cachedConfiguration.getSchemaMediator();
+        RealmProxyMediator schema = newConfiguration.getSchemaMediator();
+        if (!cachedSchema.equals(schema)) {
+            throw new IllegalArgumentException("Two configurations with different schemas are trying to open " +
+                    "the same Realm file. Their schema must be the same: " + newConfiguration.getPath());
+        }
+
+        // Check if the durability is the same
+        SharedGroup.Durability cachedDurability = cachedConfiguration.getDurability();
+        SharedGroup.Durability newDurability = newConfiguration.getDurability();
+        if (!cachedDurability.equals(newDurability)) {
+            throw new IllegalArgumentException("A Realm cannot be both in-memory and persisted. Two conflicting " +
+                    "configurations pointing to " + newConfiguration.getPath() + " are being used.");
+        }
+
+        // Should never get here
+        throw new IllegalArgumentException(
+                "The new configuration pointing to " + newConfiguration.getPath() +
+                        " has a different hash code with a cached configuration. " +
+                        "The new configuration's hashcode: " + newConfiguration.hashCode() +
+                        "The cached configuration's hashcode: " + cachedConfiguration.hashCode() + ".");
+    }
+
+    /**
+     * Runs the callback function with the total reference count of {@link Realm} and {@link DynamicRealm} refer to the
+     * given {@link RealmConfiguration}.
+     *
+     * @param configuration the {@link RealmConfiguration} of {@link Realm} and {@link DynamicRealm}.
+     * @param callback the callback will be executed with the global reference count.
+     *
+     */
+    static synchronized void getGlobalRefCount(RealmConfiguration configuration, GlobalReferenceCountCallback callback) {
```

Comment:
Something about the name of this method and the callback class just feels wrong to me.
Isn't the purpose of this method to run some code using the RealmCache, and getting the ref counter is just a side-effect of that?

Not sure I have a good alternative at the moment though